cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationEnforcementService.java

```diff
@@ -0,0 +1,159 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.security.authorization;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.proto.security.Privilege;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * A {@link AuthorizationEnforcer} used to enforce authorization policies in programs. If caching is enabled using
+ * {@link co.cask.cdap.common.conf.Constants.Security.Authorization#CACHE_ENABLED}, authorization policies are cached
+ * locally. A thread refreshes the cached policies periodically.
+ */
+@Singleton
+public class AuthorizationEnforcementService extends AbstractScheduledService implements AuthorizationEnforcer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthorizationEnforcementService.class);
+
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final boolean authorizationEnabled;
+  private final boolean cacheEnabled;
+  private final int cacheTtlSeconds;
+  private final Map<Principal, Set<Privilege>> authPolicyCache;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  @Inject
+  public AuthorizationEnforcementService(AuthorizerInstantiator authorizerInstantiator, CConfiguration cConf) {
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    this.cacheTtlSeconds = cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS);
+    this.authPolicyCache = new HashMap<>();
+    ReadWriteLock lock = new ReentrantReadWriteLock();
+    this.readLock = lock.readLock();
+    this.writeLock = lock.writeLock();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheTtlSeconds, TimeUnit.SECONDS);
```

Comment:
So the TTL is not exactly a TTL. This is because it takes time to update the cache and the next update time is actually the previous update finish time + cacheTTL. If we want to call it ttl, we should have

1. Start updating the cache earlier then the TTL
2. Have the cache expires automatically when TTL is reached